File: src/main/java/eu/mshade/enderframe/wrapper/Wrapper.java
package eu.mshade.enderframe.wrapper;

import java.util.HashMap;
import java.util.Map;

public abstract class Wrapper<I, O> {

    protected Map<I, O> outputByInput = new HashMap<>();
    protected Map<O, I> inputByOutput = new HashMap<>();

    public O wrap(I input) {
        return this.outputByInput.get(input);
    }

    public I reverse(O output) {
        return this.inputByOutput.get(output);
    }

    public void register(I i, O o) {
        this.inputByOutput.put(o, i);
        this.outputByInput.put(i, o);
    }

    public void registerInput(I i, O o) {
        this.outputByInput.put(i, o);
    }

    public void registerOutput(O o, I i) {
        this.inputByOutput.put(o, i);
    }

    public boolean isSupport(I i) {
        return this.outputByInput.containsKey(i);
    }

}

File: src/main/java/eu/mshade/enderframe/wrapper/Wrapper.kt
package eu.mshade.enderframe.wrapper

abstract class Wrapper<I, O>: Mapper<I, O> {

    private val outputByInput: MutableMap<I, O> = mutableMapOf()
    private val inputByOutput: MutableMap<O, I> = mutableMapOf()

    fun map(input: I): O? = outputByInput[input]

    fun reverseMap(output: O): I? = inputByOutput[output]

    fun register(input: I, output: O) {
        inputByOutput[output] = input
        outputByInput[input] = output
    }

    fun registerMapping(input: I, output: O) {
        outputByInput[input] = output
    }

    fun registerReverseMapping(input: I, output: O) {
        inputByOutput[output] = input
    }
}